C# Implementation of Add Method In External Editor for VS.

magian1127 opened the pull request:
After reviewing the source code for a while, I couldn't figure out the appropriate place to put the method. Therefore, I ended up creating a new class to store it.

RedworkDE commented on modules/mono/editor/GodotTools/GodotTools.OpenVisualStudio/Program.cs:
```diff
@@ -44,6 +44,21 @@ private static int Main(string[] args)
+            if (dte == null && args.Length > 1 && args[1].Contains(";AddMethod"))
+            {
+                var stopwatch = new System.Diagnostics.Stopwatch();
+                stopwatch.Start();
```
No need to use stopwatch here, just `DateTime.Now` and subtraction is enough precision; alternatively if you just do it for the code simplicity use the static `Stopwatch.StartNew()` method and remove the `Stop()` call (stopwatch does have any active components that actually get stopped)

magian1127 replied:
Changed to `DateTime.Now`

AThousandShips commented on modules/mono/editor/GodotTools/GodotTools/GodotSharpEditor.cs:
```diff
@@ -388,6 +388,45 @@ public Error OpenInExternalEditor(Script script, int line, int col)
         }
 
+        [UsedImplicitly]
+        public void AddMethodInExternalEditor(Script script, string function, string[] parameters)
```
```suggestion
        public void AddMethodInExternalEditor(Script script, string method, string[] parameters)
```

AThousandShips commented on modules/mono/language_server/csharp_language_server.cpp:
```diff
@@ -0,0 +1,63 @@
+}
+
+void CSharpLanguageServer::_bind_methods() {
+	ClassDB::bind_method(D_METHOD("add_method_in_external_editor"), &CSharpLanguageServer::add_method_in_external_editor);
```
```suggestion
	ClassDB::bind_method(D_METHOD("add_method_in_external_editor", "obj", "method", "args"), &CSharpLanguageServer::add_method_in_external_editor);
```
You need to add the arguments to the method.

magian1127 commented on modules/mono/language_server/csharp_language_server.h:
```diff
@@ -0,0 +1,49 @@
+
+protected:
+	static void _bind_methods();
+	void add_method_in_external_editor(Object *obj, String function, PackedStringArray args);
```
I copied the parameter from somewhere else, I didn't pay attention to the name, I'll pay more attention next time.

AThousandShips commented on modules/mono/SCsub:
```diff
@@ -15,6 +15,7 @@ mono_configure.configure(env, env_mono)
 
 env_mono.add_source_files(env.modules_sources, "*.cpp")
 env_mono.add_source_files(env.modules_sources, "glue/*.cpp")
+env_mono.add_source_files(env.modules_sources, "language_server/*.cpp")
```
This must be moved below with `editor/*.cpp`

magian1127 commented on modules/mono/language_server/csharp_language_server.cpp:
```diff
@@ -0,0 +1,63 @@
+void CSharpLanguageServer::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_ENTER_TREE: {
+			EditorNode *editor_node = EditorNode::get_singleton();
+			editor_node->connect("script_add_function_request", callable_mp(this, &CSharpLanguageServer::add_method_in_external_editor));
```
I've tried it, but I'm not sure why it's giving me an error. Then I created a new class.

magian1127 commented on modules/mono/register_types.cpp:
```diff
@@ -34,12 +34,24 @@
 
+#ifdef TOOLS_ENABLED
+static void _editor_init() {
+	CSharpLanguageServer *csharp_plugin = memnew(CSharpLanguageServer);
+	EditorNode::get_singleton()->add_editor_plugin(csharp_plugin);
```
I put it at the end of the `csharp_script.cpp _editor_init_callback` method, but in Godot's actual use, it generated a parent-child related error, so I put it in `register_types`.

kastromatos commented:
Should this changes fix #41283 issue? At least for VS?

AThousandShips commented on modules/mono/language_server/csharp_language_server.cpp:
```diff
@@ -0,0 +1,63 @@
+			EditorNode *editor_node = EditorNode::get_singleton();
+			editor_node->connect("script_add_function_request", callable_mp(this, &CSharpLanguageServer::add_method_in_external_editor));
+		} break;
+		case NOTIFICATION_EXIT_TREE: {
+		} break;
+		case NOTIFICATION_INTERNAL_PROCESS: {
+		} break;
+		case EditorSettings::NOTIFICATION_EDITOR_SETTINGS_CHANGED: {
+		} break;
```
```suggestion
```
Unnecessary here

magian1127 commented:
The solutions I have tried are as follows:

In the `GodotSharpEditor.cs` file, I cannot access the `EditorNode` (`EditorNode::get_singleton()` in C++), so I cannot connect to the `script_add_function_request` signal. Unless the `EditorNode` is exposed for C# usage, it is not possible to write the method in the `GodotSharpEditor.cs` file.

In the `csharp_script`, adding the new method only works when placed in the `CSharpLanguage` class. It is not suitable for the `CSharpScript` and other classes. I added `_bind_methods` to the `CSharpLanguage` class and bound the `script_add_function_request` signal in the `_editor_init_callback`. Although it compiles successfully, it crashes upon execution (it is speculated that `CSharpLanguage` does not execute `_bind_methods`).

In the end, I had to create a new class to implement it. Initially, I tried adding `add_editor_plugin` at the end of the `_editor_init_callback` in the `CSharpLanguage` class, but it resulted in an error during the usage of the new method (I don't remember the specific error, but it was related to parent-child relationships). Therefore, I moved it to "register_types.cpp".

magian1127 committed:
C# Implementation of Add Method In External Editor for VS.
